File: Ejercicios_lvl1/CoachManager.py
# Por Felipe (ZenTial)
# 1. Manchester United FC has hired you as a developer. Develop a program that helps the coach identify their fastest player, player with the most goals, assists, passing accuracy, and defensive involvements.
# The program functions as follows: The coach accesses the system and encounters a menu with the following options:
# Player Review: By entering the player's jersey number, they can access the player's characteristics.
# Compare two players: The system prompts for two jersey numbers and displays the data of both players on screen.
# Identify the fastest player: Displays the player with the most points in speed.
# Identify the top goalscorer: Displays the player with the most points in goals.
# Identify the player with the most assists: Displays the player with the most points in assists.
# Identify the player with the highest passing accuracy: Displays the player with the most points in passing accuracy.
# Identify the player with the most defensive involvements: Displays the player with the most points in defensive involvements.
# The system should also allow returning to the main menu.
from os import system

# Declaración de variables

camiseta_8 = {'nombre': 'Bruno Fernandes', 'goles': 5, 'puntos velocidad': 6,
              'puntos asistencia': 9, 'puntos precision pase': 10, 'participacion defensiva': 3}
camiseta_11 = {'nombre': 'Rasmus Hojlund', 'goles': 12, 'puntos velocidad': 8,
               'puntos asistencia': 2, 'puntos precision pase': 6, 'participacion defensiva': 2}
camiseta_5 = {'nombre': 'Harry Maguire', 'goles': 1, 'puntos velocidad': 5,
              'puntos asistencia': 1, 'puntos precision pase': 7, 'participacion defensiva': 9}
camiseta_17 = {'nombre': 'Alejandro Garnacho', 'goles': 8, 'puntos velocidad': 7,
               'puntos asistencia': 8, 'puntos precision pase': 16, 'participacion defensiva': 0}
camiseta_7 = {'nombre': 'Mason Mount', 'goles': 2, 'puntos velocidad': 6,
              'puntos asistencia': 4, 'puntos precision pase': 8, 'participacion defensiva': 1}
numeros_camisetas = ['8', '11', '5', '17', '7']
lista_camisetas = [camiseta_8, camiseta_11, camiseta_5, camiseta_17, camiseta_7]


# Función jugador más velocidad
def jugador_masrapido():
    velocidades = []
    for jugador in lista_camisetas:
        velocidades.append(jugador['puntos velocidad'])
    mas_rapido = max(velocidades)
    indice = velocidades.index(mas_rapido)
    print(f'El jugador más rapido es {lista_camisetas[indice]['nombre']}')
    print(f'Su velocidad es: {mas_rapido}')
    input('Presione cualquier tecla para volver al menú')


# Función jugador con más goles
def jugador_masgoles():
    goles = []
    for jugador in lista_camisetas:
        goles.append(jugador['goles'])
    mas_goles = max(goles)
    indice = goles.index(mas_goles)
    print(f'El jugador con más goles es {lista_camisetas[indice]['nombre']}')
    print(f'Su cantidad de goles es: {mas_goles}')
    input('Presione cualquier tecla para volver al menú')


# Función jugador con más asistencias
def jugador_masasistencias():
    asistencias = []
    for jugador in lista_camisetas:
        asistencias.append(jugador['puntos asistencia'])
    mas_asistencias = max(asistencias)
    indice = asistencias.index(mas_asistencias)
    print(f'El jugador con más asistencias es {lista_camisetas[indice]['nombre']}')
    print(f'Su cantidad de asistencias es: {mas_asistencias}')
    input('Presione cualquier tecla para volver al menú')


# Función jugador con más precisión de pase
def jugador_precisionpase():
    pases = []
    for jugador in lista_camisetas:
        pases.append(jugador['puntos precision pase'])
    mas_pases = max(pases)
    indice = pases.index(mas_pases)
    print(f'El jugador con más precisión de pase es {lista_camisetas[indice]['nombre']}')
    print(f'Su puntaje en precisión de pases es: {mas_pases}')
    input('Presione cualquier tecla para volver al menú')


# Función jugador con más participación defensiva
def jugador_defensivo():
    defensas = []
    for jugador in lista_camisetas:
        defensas.append(jugador['participacion defensiva'])
    mas_defensas = max(defensas)
    indice = defensas.index(mas_defensas)
    print(f'El jugador con mayor participación defensiva es {lista_camisetas[indice]['nombre']}')
    print(f'Su puntaje en participación defensiva es: {mas_defensas}')
    input('Presione cualquier tecla para volver al menú')


# Función de revisión del jugador
def revision_jugador():
    while True:
        print('*' * 50)
        print('Numeros de camiseta de jugadores:')
        print(', '.join(numeros_camisetas))
        try:
            accion = input('Elige un número de camiseta: ')
            indice = numeros_camisetas.index(accion)
            datos = lista_camisetas[indice]
        except ValueError:
            system('cls')
            print('Error, elige una opción correcta')
        else:
            system('cls')
            for key, value in datos.items():
                print(f'{key.capitalize()}: {value}')
            input('Presione cualquier tecla para volver al menu')
            return


# Función comparación
def comparacion():
    while True:
        print('*' * 50)
        print('Elige a 2 jugadores a comparar')
        print('Numeros de camiseta de jugadores:')
        print(', '.join(numeros_camisetas))
        try:
            accion1 = input('Elige al primer jugador: ')
            indice1 = numeros_camisetas.index(accion1)
            accion2 = input('Elige al segundo jugador: ')
            indice2 = numeros_camisetas.index(accion2)
        except ValueError:
            system('cls')
            print('Error, ingrese los valores correctos')
        else:
            system('cls')
            datos_jugador1 = lista_camisetas[indice1]
            datos_jugador2 = lista_camisetas[indice2]
            for key, value in datos_jugador1.items():
                print(f'{key.capitalize()} : {value}')
            print('')
            for key, value in datos_jugador2.items():
                print(f'{key.capitalize()}: {value}')
            input('Presione cualquier tecla para volver al menu')
            return


# Función menú
def menu():
    while True:
        print('*' * 50)
        print('Bienvenido al programa de jugadores del Manchester United')
        print('[1] Ver datos de jugador\n[2] Comparar 2 jugadores\n[3] Ver jugador con más goles\n'
              '[4] Ver jugador con más velocidad\n[5] Ver jugador con más asistencias\n'
              '[6] Ver jugador con más precisión de pase\n[7] Ver jugador más defensivo\n[8] Terminar programa')
        try:
            accion = input('Elige una opción: ')
            ['1', '2', '3', '4', '5', '6', '7', '8'].index(accion)
        except ValueError:
            system('cls')
            print('Error, Elige una opción valida')
        else:
            return accion


# Función main
def main():
    validacion = True
    while validacion:
        system('cls')
        accion = menu()
        if accion == '1':
            system('cls')
            revision_jugador()
            continue
        elif accion == '2':
            system('cls')
            comparacion()
            continue

        elif accion == '3':
            system('cls')
            jugador_masgoles()
            continue

        elif accion == '4':
            system('cls')
            jugador_masrapido()
            continue

        elif accion == '5':
            system('cls')
            jugador_masasistencias()
            continue

        elif accion == '6':
            system('cls')
            jugador_precisionpase()
            continue

        elif accion == '7':
            system('cls')
            jugador_defensivo()
            continue

        else:
            system('cls')
            print('El programa ha sido finalizado')
            validacion = False


if __name__ == '__main__':
    main()

File: Reto_1/coachManager.js
// 1. Manchester United FC has hired you as a developer. Develop a program that helps the coach identify their fastest player, player with the most goals, assists, passing accuracy, and defensive involvements.
// The system should also allow comparison between two players. Use the following player profiles:

// Bruno Fernandes: 5 goals, 6 points in speed, 9 points in assists, 10 points in passing accuracy, 3 defensive involvements. Corresponds to jersey number 8.
// Rasmus Hojlund: 12 goals, 8 points in speed, 2 points in assists, 6 points in passing accuracy, 2 defensive involvements. Corresponds to jersey number 11.
// Harry Maguire: 1 goal, 5 points in speed, 1 point in assists, 7 points in passing accuracy, 9 defensive involvements. Corresponds to jersey number 5.
// Alejandro Garnacho: 8 goals, 7 points in speed, 8 points in assists, 6 points in passing accuracy, 0 defensive involvements. Corresponds to jersey number 17.
// Mason Mount: 2 goals, 6 points in speed, 4 points in assists, 8 points in passing accuracy, 1 defensive involvement. Corresponds to jersey number 7.
// The program functions as follows: The coach accesses the system and encounters a menu with the following options:

// Player Review: By entering the player's jersey number, they can access the player's characteristics.
// Compare two players: The system prompts for two jersey numbers and displays the data of both players on screen.
// Identify the fastest player: Displays the player with the most points in speed.
// Identify the top goal scorer: Displays the player with the most points in goals.
// Identify the player with the most assists: Displays the player with the most points in assists.
// Identify the player with the highest passing accuracy: Displays the player with the most points in passing accuracy.
// Identify the player with the most defensive involvements: Displays the player with the most points in defensive involvements.
// The system should also allow returning to the main menu.
import * as readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";
import { log } from "node:console";
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
const players = [
  {
    name: "Bruno Fernandes",
    goals: 5,
    speed: 6,
    assists: 9,
    passingAccuracy: 10,
    defensiveInvolvements: 3,
    jerseyNumber: 8,
  },
  {
    name: "Rasmus Hojlund",
    goals: 12,
    speed: 8,
    assists: 2,
    passingAccuracy: 6,
    defensiveInvolvements: 2,
    jerseyNumber: 11,
  },
  {
    name: "Harry Maguire",
    goals: 1,
    speed: 5,
    assists: 1,
    passingAccuracy: 7,
    defensiveInvolvements: 9,
    jerseyNumber: 5,
  },
  {
    name: "Alejandro Garnacho",
    goals: 8,
    speed: 7,
    assists: 8,
    passingAccuracy: 6,
    defensiveInvolvements: 0,
    jerseyNumber: 17,
  },
  {
    name: "Mason Mount",
    goals: 2,
    speed: 6,
    assists: 4,
    passingAccuracy: 8,
    defensiveInvolvements: 1,
    jerseyNumber: 7,
  },
];
async function askToContinue() {
  const answer = await rl.question(
    "Do you want to continue in the program? (y/n): "
  );
  if (answer.toLowerCase() !== "y") {
    isProgramRunning = false;
  }
}

async function playerReview() {
  const jerseyNumber = await rl.question(
    "Input the jersey number of the player: "
  );
  const player = players.find((player) => player.jerseyNumber == jerseyNumber);
  if (player) {
    console.log("Statistics");
    console.log(`
Name: ${player.name}
Number of goals: ${player.goals}    
Points in speed: ${player.speed}    
Points in assists: ${player.assists}    
Points in passing accuracy: ${player.passingAccuracy}    
Defensive involvements: ${player.defensiveInvolvements}    
    `);
  } else {
    console.log("Player no found");
  }
}
async function playersCompare() {
  await playerReview();
  await playerReview();
}
async function selectBestPlayer(statistic) {
  let bestPlayer = null
  players.forEach((player) => {
    if (statistic in player) {
      if (bestPlayer === null ||bestPlayer[statistic] < player[statistic]) {
        bestPlayer = player
      }
    }
  });
  return bestPlayer;
}
let isProgramRunning = true;
while (isProgramRunning) {
  console.log("Welcome to the Player Review of Manchester United FC!");
  console.log("1. Player Review:");
  console.log("2. Compare two players:");
  console.log("3. Identify the fastest player");
  console.log("4. Identify the top goal scorer");
  console.log("5. Identify the player with the most assists");
  console.log("6. Identify the player with the highest passing accuracy");
  console.log("7. Identify the player with the most defensive involvements");
  console.log("8. Exit");
  let optionMenu = await rl.question(
    "select the option according to the number in the menu: "
  );
  switch (Number(optionMenu)) {
    case 1:
      await playerReview();
      await askToContinue();
      break;
    case 2:
      await playersCompare();
      await askToContinue();
      break;
    case 3:
      const fastestPlayer = await selectBestPlayer("speed");
      console.log(
        `Fastest player: ${fastestPlayer.name} with a speed score of ${fastestPlayer.speed}`
      );
      await askToContinue();

      break;
    case 4:
      const topGoalScorer = await selectBestPlayer("goals");
      console.log(
        `Top goal scorer: ${topGoalScorer.name} with ${topGoalScorer.goals} goals`
      );
      await askToContinue();

      break;
    case 5:
      const mostAssistsPlayer = await selectBestPlayer("assists");
      console.log(
        `Player with most assists: ${mostAssistsPlayer.name} with ${mostAssistsPlayer.assists} assists`
      );
      await askToContinue();
      break;
    case 6:
      const highestPassingAccuracyPlayer = await selectBestPlayer(
        "passingAccuracy"
      );
      console.log(
        `Player with highest passing accuracy: ${highestPassingAccuracyPlayer.name} with ${highestPassingAccuracyPlayer.passingAccuracy}% passing accuracy`
      );
      await askToContinue();
      break;
    case 7:
      const mostDefensivePlayer = await selectBestPlayer(
        "defensiveInvolvements"
      );
      console.log(
        `Player with most defensive involvements: ${mostDefensivePlayer.name} with ${mostDefensivePlayer.defensiveInvolvements} defensive involvements`
      );
      await askToContinue();
      break;
    case 8:
      isProgramRunning = false;
      break;

    default:
      console.log(`
Please enter a valid option.
`);
      break;
  }
}
rl.close();
